File: broker/src/main/scala/io/octopus/scala/factory/FastThreadLocalRunnable.scala
package io.octopus.scala.factory

import io.netty.util.concurrent.FastThreadLocal

/**
 * @author chenxu
 * @version 1
 */

class FastThreadLocalRunnable(runnable: Runnable) extends Runnable {

  override def run(): Unit = {
    try this.runnable.run()
    finally FastThreadLocal.removeAll()
  }

}

object FastThreadLocalRunnable {
  def wrap(runnable: Runnable) = if (runnable.isInstanceOf[FastThreadLocalRunnable]) runnable
  else new FastThreadLocalRunnable(runnable)
}

File: broker-plugs/broker-mqtt/src/main/java/io/octopus/factory/FastThreadLocalRunnable.java
//package io.octopus.factory;
//
//import io.netty.util.concurrent.FastThreadLocal;
//import io.netty.util.internal.ObjectUtil;
//
///**
// * @author chenxu
// * @version 1
// */
//final class FastThreadLocalRunnable implements Runnable {
//    private final Runnable runnable;
//
//    private FastThreadLocalRunnable(Runnable runnable) {
//        this.runnable = (Runnable) ObjectUtil.checkNotNull(runnable, "runnable");
//    }
//
//    public void run() {
//        try {
//            this.runnable.run();
//        } finally {
//            FastThreadLocal.removeAll();
//        }
//
//    }
//
//    static Runnable wrap(Runnable runnable) {
//        return (Runnable)(runnable instanceof FastThreadLocalRunnable ? runnable : new FastThreadLocalRunnable(runnable));
//    }
//}